mungegithub/mungers/approval-handler.go

```diff
@@ -0,0 +1,146 @@
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+)
+
+// ApprovalHandler will try to add "approved" label once
+// all files of change has been approved by approvers.
+type ApprovalHandler struct {
+	features *features.Features
+}
+
+func init() {
+	h := &ApprovalHandler{}
+	RegisterMungerOrDie(h)
+}
+
+// Name is the name usable in --pr-mungers
+func (*ApprovalHandler) Name() string { return "approval-handler" }
+
+// RequiredFeatures is a slice of 'features' that must be provided
+func (*ApprovalHandler) RequiredFeatures() []string {
+	return []string{features.RepoFeatureName, features.AliasesFeature}
+}
+
+// Initialize will initialize the munger
+func (h *ApprovalHandler) Initialize(config *github.Config, features *features.Features) error {
+	h.features = features
+	return nil
+}
+
+// EachLoop is called at the start of every munge loop
+func (*ApprovalHandler) EachLoop() error { return nil }
+
+// AddFlags will add any request flags to the cobra `cmd`
+func (*ApprovalHandler) AddFlags(cmd *cobra.Command, config *github.Config) {}
+
+// Munge is the workhorse the will actually make updates to the PR
+// The algorithm goes as:
+// - Initially, we set up approverMap
+//   - Go through all comments after latest commit. If any approver said "/approve", add him to approverMap.
+// - For each file, we see if any approver of this file is in approverMap.
+//   - An approver of a file is defined as:
+//     - It's known that each dir has a list of approvers. (This might not hold true. For usability, current situation is enough.)
+//     - Approver of a dir is also the approver of child dirs.
+//   - We look at only top 3 dir level's approvers .
+// - Iff all files has been approved, the bot will add "approved" label.
+func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	if !obj.IsPR() {
+		return
+	}
+	files, err := obj.ListFiles()
+	if err != nil {
+		glog.Errorf("failed to list files in this PR: %v", err)
+		return
+	}
+
+	comments, err := getCommentsAfterLastModified(obj)
+	if err != nil {
+		glog.Errorf("failed to get comments in this PR: %v", err)
+		return
+	}
+
+	approverMap := sets.String{}
+
+	// from oldest to latest
+	for i := len(comments) - 1; i >= 0; i-- {
+		c := comments[i]
+
+		if !mungerutil.IsValidUser(c.User) {
+			continue
+		}
+
+		fields := strings.Fields(strings.TrimSpace(*c.Body))
+
+		if len(fields) == 1 && strings.ToLower(fields[0]) == "/approve" {
+			approverMap.Insert(*c.User.Login)
+			continue
+		}
+
+		if len(fields) == 2 && strings.ToLower(fields[0]) == "/approve" && strings.ToLower(fields[1]) == "cancel" {
+			approverMap.Delete(*c.User.Login)
+		}
+	}
+
+	for _, file := range files {
+		if !h.hasApproval(*file.Filename, approverMap) {
+			return
+		}
+	}
+	obj.AddLabel(approvedLabel)
+}
+
+func (h *ApprovalHandler) hasApproval(filename string, approverMap sets.String) bool {
+	paths := strings.Split(filename, "/")
+	var p string
+	for i := 0; i < len(paths) && i < 3; i++ {
```

Comment:
can we make 3 a parameter so it's clear that it represents the maxDepth of the search?